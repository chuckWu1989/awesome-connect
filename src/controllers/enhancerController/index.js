/* eslint no-underscore-dangle: off */
import uuidv4 from 'uuid/v4';

export const getId = () => uuidv4();
export const getName = (id, name) => {
  const result = name !== undefined ? name : id;
  return result;
};
export function mergeProps(props, name, store) {
  return (
    '_reactInternalInstance' in this &&
    '_context' in this._reactInternalInstance &&
    typeof this._reactInternalInstance._context === 'object' &&
    !('store' in this._reactInternalInstance._context) ?
      { ...props, name, store } :
      { ...props, name }
  );
}
