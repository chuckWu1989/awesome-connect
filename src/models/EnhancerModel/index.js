import { DefineProps, DefaultValue, Type, BaseModel } from 'model-decorator';
import PropTypes from 'prop-types';

class EnhancerModel extends BaseModel {
  @DefineProps()
  @Type(PropTypes.oneOfType([PropTypes.string, PropTypes.number]))
  name

  @DefineProps()
  @Type(PropTypes.func)
  @DefaultValue(() => {})
  getId

  @DefineProps()
  @Type(PropTypes.func)
  @DefaultValue(() => {})
  getName

  @DefineProps()
  @Type(PropTypes.func)
  @DefaultValue(() => {})
  mergeProps
}

export default EnhancerModel;
