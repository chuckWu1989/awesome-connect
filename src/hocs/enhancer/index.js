/* eslint function-paren-newline: ["error", "consistent"] */
import React, { Component } from 'react';
import { ViewModel } from 'model-decorator';
import EnhancerModel from '../../models/EnhancerModel';

const enhancer = (WrappedComponent, store) => (
  ViewModel(EnhancerModel)(
    class extends Component {
      constructor(props) {
        super(props);
        this.state = { id: undefined };
      }
      componentWillMount() {
        const { props: { model } } = this;
        this.setState({ id: model.getId.val() });
      }
      render() {
        const { state: { id }, props: { model, ...otherProps } } = this;
        const { name, getName, mergeProps } = model.getValues();
        const componentName = getName(id, name);
        const ownProps = mergeProps.call(this, otherProps, componentName, store);
        return (
          <WrappedComponent {...ownProps} />
        );
      }
    },
  )
);

export default enhancer;
