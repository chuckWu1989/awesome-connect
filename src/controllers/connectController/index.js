import lodash from 'lodash';
import { connect as reduxConnect } from 'react-redux';
import enhancerRouter from '../../routers/enhancerRouter';

export const getMapStateToProps = (mapStateToProps, defaultProps) => (
  typeof mapStateToProps === 'function' ?
    (state, props) => (
      mapStateToProps(state, lodash.merge({}, defaultProps, props))
    ) :
    null
);
export const getMapDispatchToProps = (mapDispatchToProps, defaultProps) => (
  typeof mapDispatchToProps === 'function' ?
    (dispatch, props) => (
      mapDispatchToProps(dispatch, lodash.merge({}, defaultProps, props))
    ) :
    null
);
export const getMergeProps = (mergeProps, defaultProps) => (
  typeof mergeProps === 'function' ?
    (stateProps, dispatchProps, props) => (
      mergeProps(
        stateProps, dispatchProps, lodash.merge({}, defaultProps, props),
      )
    ) :
    null
);
export const connect = store => (
  (mapStateToProps = null, mapDispatchToProps = null, mergeProps = null) => (
    (component, type = {}) => {
      if (typeof component !== 'function') {
        const error = new TypeError('The component must be a function', 'connect');
        if (process.env.NODE_ENV !== 'production') {
          console.error(error.toString());
        }
        throw error;
      }
      const { defaultProps = {} } = component;
      const newMapStateToProps = getMapStateToProps(mapStateToProps, defaultProps);
      const newMapDispatchToProps = getMapDispatchToProps(mapDispatchToProps, defaultProps);
      const newMergeProps = getMergeProps(mergeProps, defaultProps);
      const ConnectComponent = reduxConnect(
        newMapStateToProps,
        newMapDispatchToProps,
        newMergeProps,
      )(component);
      const HOC = enhancerRouter(ConnectComponent, store);
      Object.keys(type).forEach((key) => {
        HOC[key] = type[key];
      });
      return HOC;
    }
  )
);
