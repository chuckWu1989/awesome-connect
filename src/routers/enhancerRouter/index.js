import router from '../../hocs/router';
import { getId, getName, mergeProps } from '../../controllers/enhancerController';
import enhancer from '../../hocs/enhancer';

const method = {
  getId,
  getName,
  mergeProps,
};

export default (WrappedComponent, store) => router(method)(enhancer(WrappedComponent, store));
