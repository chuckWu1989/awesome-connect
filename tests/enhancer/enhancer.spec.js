import React from 'react';
import toJson from 'enzyme-to-json';
import enhancer from '../../src/hocs/enhancer';
import EnhancerModel from '../../src/models/EnhancerModel';
import enhancerRouter from '../../src/routers/enhancerRouter';
import * as controller from '../../src/controllers/enhancerController';

const { shallow } = global;

describe('enhancer', () => {
  describe('model', () => {
    it('should define properties properly', () => {
      const model = new EnhancerModel();
      expect(model).toMatchSnapshot();
    });
  });
  describe('hoc', () => {
    it('should render properly', () => {
      const WrappedComponent = () => <div />;
      const store = undefined;
      const OutputComponent = enhancer(WrappedComponent, store);
      const tree = toJson(shallow(<OutputComponent />));
      expect(tree).toMatchSnapshot();
    });
    it('should call methods during lifecycle', () => {
      const WrappedComponent = () => <div />;
      const store = undefined;
      const OutputComponent = enhancer(WrappedComponent, store);
      const ownProps = {
        name: 'test',
        getId: jest.fn(() => 'test'),
        getName: jest.fn(() => 'test'),
        mergeProps: jest.fn(() => ({})),
      };
      shallow(
        <OutputComponent {...ownProps} />,
        { lifecycleExperimental: true },
      ).shallow();
      expect(ownProps.getId).toHaveBeenCalledTimes(1);
      expect(ownProps.getName).toHaveBeenCalledTimes(1);
      expect(ownProps.mergeProps).toHaveBeenCalledTimes(1);
    });
  });
  describe('router', () => {
    it('should pass correct properties', () => {
      const WrappedComponent = () => <div />;
      const store = undefined;
      const OutputComponent = enhancerRouter(WrappedComponent, store);
      const props = shallow(<OutputComponent />).props();
      expect('getId' in props).toBeTruthy();
      expect('getName' in props).toBeTruthy();
      expect('mergeProps' in props).toBeTruthy();
    });
  });
  describe('controller', () => {
    describe('getId', () => {
      it('should generate uuid', () => {
        const id = controller.getId();
        expect(typeof id).toBe('string');
      });
    });
    describe('getName', () => {
      it('should get name if name is defined', () => {
        const name = 'name';
        const id = 'id';
        const result = controller.getName(id, name);
        expect(result).toBe(name);
      });
      it('should get id if name is undefined', () => {
        const name = undefined;
        const id = 'id';
        const result = controller.getName(id, name);
        expect(result).toBe(id);
      });
    });
    describe('mergeProps', () => {
      const store = { test: '123' };
      const name = 'name';
      const props = { myProps: 'props' };
      it('should return props, name and store if _context exists in _reactInternalInstance and store is undefined', () => {
        const self = { _reactInternalInstance: { _context: {} } };
        const result = controller.mergeProps.call(self, props, name, store);
        expect(result.name).toBe(name);
        expect(result.store).toBe(store);
        expect(result.myProps).toBe(props.myProps);
      });
      it('should return props and name if _context is undefined or store is defined', () => {
        const self = { _reactInternalInstance: { _context: { store: '123' } } };
        const result = controller.mergeProps.call(self, props, name, store);
        expect(result.name).toBe(name);
        expect(result.store).toBeUndefined();
        expect(result.myProps).toBe(props.myProps);
      });
    });
  });
});
