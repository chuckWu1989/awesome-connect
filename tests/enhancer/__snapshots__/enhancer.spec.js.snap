// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enhancer hoc should render properly 1`] = `
<_class
  model={
    EnhancerModel {
      "getId": ModelItem {},
      "getName": ModelItem {},
      "mergeProps": ModelItem {},
      "name": ModelItem {},
    }
  }
/>
`;

exports[`enhancer model should define properties properly 1`] = `
EnhancerModel {
  "getId": ModelItem {},
  "getName": ModelItem {},
  "mergeProps": ModelItem {},
  "name": ModelItem {},
}
`;
